### PR TITLE
Synchronize and improve the global parallelism

### PR DESCRIPTION
**Pull request by yuchengliu1**
## Type of Change

fix bug
improve performance

## Description

fix mha accuracy on CPU with avx512 but without amx
instead omp parallel with unified parallel interface 
fix mha cache and adjust thread set in model inference
use single thread in rope next token

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Review by DDEle — APPROVED** [review]
LGTM.

**Comment by luoyu-intel** [comment]
TODO: a). perf check b). remove omp

**Comment by yuchengliu1** [comment]
performance on SPR has no change.

**Comment by yuchengliu1** [comment]
performance on 13900K (Linux)
===================before====================================
model_print_timings:        load time =  5039.76 ms
model_print_timings:      sample time =    13.87 ms /    32 runs   (    0.43 ms per token)
model_print_timings: prompt eval time =  5039.36 ms /  1024 tokens (    4.92 ms per token)
model_print_timings:        eval time =  2378.82 ms /    31 runs   (   76.74 ms per token)
model_print_timings:       total time =  7438.79 ms


===================after====================================
model_print_timings:        load time =  3938.19 ms
model_print_timings:      sample time =    15.15 ms /    32 runs   (    0.47 ms per token)
model_print_timings: prompt eval time =  3937.72 ms /  1024 tokens (    3.85 ms per token)
model_print_timings:        eval time =  2135.32 ms /    31 runs   (   68.88 ms per token)
model_print_timings:       total time =  6095.11 ms

File: neural_speed/core/layers/mha_dense.cpp [hunk 1]
```diff
@@ -63,7 +63,8 @@ void bestla_fusion_attn_int8_forward(const attn_int8_fwd_args_t* params) {
 }
 size_t bestla_fusion_attn_workspace_size(const attn_shape_t* params) {
   const auto& p = *params;  // TODO(Yi): Better way to get tmp size?
-  return size_t(ne_threading::get()->num_threads() * sizeof(float) * 16) * padto(padto(p.sl_kv, 48), 64);
+  GetCPUDevice();
+  return size_t(_cd->getThreads() * sizeof(float) * 16) * padto(padto(p.sl_kv, 48), 64);
```
Comment (luoyu-intel) [inline neural_speed/core/layers/mha_dense.cpp:67]: ne_threading::get()->num_threads() should be the actual threads? why change it to device's threads?

File: bestla/bestla/bestla_parallel.h [hunk 2]
```diff
@@ -30,6 +30,36 @@ using thread_func = std::function<void(int tid)>;
...
+
+  // equal to "for(int i=begin1;i<end1;i+=step1)"
+  void parallel_for_collapse(const int& begin1, const int& end1, const int& step1,
+                             const std::function<void(int)>& func) {
+    assert(end1 >= begin1 && step1 > 0);
+    parallel_for([&](int tidx) {
+      const int length = (end1 - begin1 + step1 - 1) / step1;
+      const int block_size = length / mThreadNum;
+      const int block_remain = length % mThreadNum;
+      const int offset = std::min(tidx, block_remain) + tidx * block_size;
```
Comment (luoyu-intel) [inline bestla/bestla/bestla_parallel.h:42]: what does this `std::min(tidx, block_remain)` mean?

Comment (yuchengliu1) [reply bestla/bestla/bestla_parallel.h:42]: The first block_remain threads will do `block_size+1` and the others will do `block_size` in the division. `offset` is the begin of each thread.

**Comment by bil-ash** [comment]
Currently, in `docs/fused_attention.md` the `Runtime ISA` is mentioned as `AMX_BF16` . I guess with this PR the text should be changed to `AVX512F or newer` .

**Comment by luoyu-intel** [comment]
> Currently, in `docs/fused_attention.md` the `Runtime ISA` is mentioned as `AMX_BF16` . I guess with this PR the text should be changed to `AVX512F or newer` .

@bil-ash  Title changed. This PR changes the global threading. `docs/fused_attention.md` is out of date, AVX512F and AVX2 are all ready for use.

**Review by luoyu-intel — APPROVED** [review]

**Pull request merged**